### PR TITLE
Imran toggle invisibility permission self and others

### PR DESCRIPTION
**Pull request by ImzIssa**
# Description
Desc

## Related PRS (if any):
This frontend PR is related to the backend PR[#906](https://github.com/OneCommunityGlobal/HGNRest/pull/906)

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
5. see video below on the rest of steps 

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/87269725/7cefb045-a9b0-46ac-ab37-7f44f9f74615

### CODE REVIEW

**Comment by ImzIssa** [comment]
> Hi, [ImzIssa](https://github.com/ImzIssa). I've tried to follow your video. At first, I've created my team, namely `KyoTestTeam`, and assigned 3 members, `KyoTest Admin`, `KyoTest Volunteer`, `KyoTest2 Volunteer`.
> 
> KyoTestTeam
> 
> As expected, when logged in as `KyoTest Admin`, I could see the `visible/invisible` toggle button, and when logged in as `KyoTest Volunteer` and `KyoTest2 volunteer`, I could not. I've logged in as `KyoTest Admin` and could successfully give permission for `Toggle Invisibility Permission Self and Others` to `KyoTest Volunteer` and `KyoTest2 Volunteer`, which is good until now.
> 
> I've made both `KyoTest Volunteer` and `KyoTest2 Volunteer` invisible by setting the toggling button to `invisible`, when I logged in as `KyoTest Volunteer` and `KyoTest2 Volunteer` separately.
> 
> KyoTest made invisible KyoTest2 made invisible
> 
> I've also made `KyoTest Admin` invisible.
> 
> KyoTest Admin invisible
> 
> Now, we expect the team members cannot see each other.
> 
> But what I have now is, I can see all members from the Dashboard no matter what account I logged in with.
> 
> When logged in as `KyoTest Admin`, I could see `KyoTest Volunteer` and `KyoTest2 Volunteer`, which could be thought as acceptable because I logged in as an administrator.
> 
> visiable logged in as admin
> 
> But when logged in as `KyoTest Volunteer`, I could also see all of the members from Dashboard, who were all set INVISIBLE.
> 
> all users visible
> 
> I've logged in as `KyoTest2 Volunteer`, and also could see all members belonging the the same team including the one who has the administrator status.
> 
> all users visible
> 
> I've logged in and out many times and went back and forth from pages to pages, and the localhost ended up with crashing, which I think from backend database error.
> 
> crash
> 
> crash log from bash
> 
> Please see the attached crash log file. [PR2183-Crash log.log](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/files/15081113/PR2183-Crash.log.log)

Hi @alla0810 
Thank you for thoroughly testing the PR, great work!. 
The dashboard was not part of the scope of my task but I agree that not visible users should not be able to see each other. I will find out if I can work on the dashboard as part of this PR or if someone is working on it already.

**Review by cgomezhub — COMMENTED** [review]
Toggle Function Issue - Admin vs. Volunteer User

During testing, we encountered an inconsistency with the toggle function:

Admin User: The toggle function works as expected, allowing admins to successfully set the volunteer user's status.

Volunteer User: However, the toggle function is not functional for volunteer users. They are unable to:

Activate the "closed eyed" option.
Change their status to "inactive."

Screenshot 2024-04-29 121317
Screenshot 2024-04-29 143911
Screenshot 2024-04-29 143935
Screenshot 2024-04-29 144353

File: src/actions/userManagement.js [hunk 1]
```diff
@@ -72,6 +72,24 @@ export const updateRehireableStatus = (user, isRehireable) => {
...
+ * @param{boolean} isVisible - the new visiblity status
+ */
+export const toggleVisibility = (user, isVisible) => {
+  const userProfile = { ...user };
+  userProfile.isVisible = isVisible
+  const requestData = { isVisible };
+  
+  const toggleVisibilityPromise = axios.patch(ENDPOINTS.TOGGLE_VISIBILITY(user._id), requestData)
+  return async dispatch => {
+    toggleVisibilityPromise.then(res => {
```
Comment (mSharifHub) [inline src/actions/userManagement.js:87]: the code is missing error handling. You are only handling success cases of axios.patch requests. However, it does not account for situations where the request may fail due to network issues, server errors, or other reasons. 
```suggestion
   .catch((err)=>{
   console.error("failed to toggle visibility:',error):
   dispatch(userProfileErrorAction(error));
   }
   }
   ........
```

**Review by mSharifHub — CHANGES_REQUESTED** [review]
After reviewing your code I noticed that you not accounting for failure due to network. In addition, on userManagement.js
`export const deleteUser = (user, option) => {
  const requestData = { option, userId: user._id };
  const deleteProfilePromise = axios.delete(ENDPOINTS.USER_PROFILE(user._id), {
    data: requestData,
  });
  return async dispatch => {
    deleteProfilePromise
      .then(res => {
        dispatch(userProfileDeleteAction(user));
      })
      .catch(err => {
        dispatch(userProfileDeleteAction(user));
      });
  };
};`

 the function dispatched the same action (userProfileDeleteAction) whether succeeds or fails. This may not allow the UI to respond appropriately.

**Review by KaushikMreddy — COMMENTED** [review]
Hi, 
I've reviewed your changes, and it works for myself ( Owner Account ) and I tried to set it for someone else as well.
image
image


I encountered an error with a value called count while opening another user account from the homepage. I'm not sure if this is related to this feature, but please take a look if possible

Click on the underlined user
image

Got this error
image

Please note that this error is only occurring for the the user: 0Eduardo Volunteer

**Comment by ImzIssa** [comment]
@cgomezhub Thank you for testing. The Leaderboard was not part of the scope for this PR

**Comment by ImzIssa** [comment]
It's not ralated but did you try clearing the cache?

**Review by nathcarnevalli — COMMENTED** [review]
I've got an error SyntaxError as shown below

image

**Comment by ImzIssa** [comment]
@nathcarnevalli Fixed. Please Test again

**Comment by ImzIssa** [comment]
@nathcarnevalli this PR does not cover the dashboard but did you try a hard reload of the page

**Comment by nathcarnevalli** [comment]
> @nathcarnevalli this PR does not cover the dashboard but did you try a hard reload of the page

Yes, but I tested now on Google Chrome and it worked. I'll approve your PR.

**Review by nathcarnevalli — APPROVED** [review]

**Review by jayyy-s — DISMISSED** [review]
Appears to be working correctly. The position automatically exists for an owner level user, and I was able use the toggle/confirm modal. I am able to assign the permission to a volunteer user and this video also shows that a mentor role does not automatically have the permission. It appears to address everything that was in scope, so I'm approving.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/71467346/cc9bacaa-2531-4838-85fa-477380181671

**Comment by nishithaShetty** [comment]
Hi, I've reviewed your changes, and it works for Admin and volunteer user.
I encountered this error, loading follow up data when I log into any account
PR#2183_1
PR#2183_2
PR#2183_3
PR#2183_4
PR#2183_error
PR#2183_error_followup_data

**Comment by Sandhya1236** [comment]
> Member
> ###
> @ImzIssa ** [ImzIssa](/ImzIssa) **
I'm unable to play the video nor download it for testing. Can u help me?

**Review by Infinite-Turkey — CHANGES_REQUESTED** [review]
image
I have followed your video, but met 1 errors: 1. Loading follow up data when I logged into my account

**Review by Kavil-Jain-514 — APPROVED** [review]
The functionality works as expected.
The test user doesn't have the visibility toggle before permission
image
Setting the permission
image
Toggle option is visible
image
User is able to hide his visibility
image


Great work!!

**Review by JatinAgrawal94 — DISMISSED** [review]
Functionality works as described. Consider a couple of warnings being shown in the console. Rest everything works fine.
Good job!



https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/47699187/8f64082d-d073-4bb8-9675-e2fd33885ebc

**Comment by akshit0211** [comment]
Hi Imran, after granting permission to a volunteer user, I can successfully toggle their invisibility or visibility.

admin user
volunteer user
Permission
Volunteer with permission

**Review by akshit0211 — DISMISSED** [review]

**Review by DiyaWadhwani — CHANGES_REQUESTED** [review]
Hi Imran,
Tried testing this as per your instructions but I'm unable to view the user profile. Reloading the browser displays the user profile for a second and the entire UI becomes blank again. 
There are around 3 API calls on the backend throwing a 404 error and my assumption is that the other API calls are dependent on either of them. (I'm guessing it's this one 
http://localhost:4500/api/notification/unread/user/66725af292f78117649bdc5b)
I have attached a video of the issue I'm facing while testing. Kindly review this.

Thank you

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/54880903/7fd0cea3-c0d1-4777-aee3-431bf2f635d4

**Review by vishavk1992 — CHANGES_REQUESTED** [review]
I am unable to start localhost:3000 due to errors and failed test suites. I have followed the steps you mentioned, but I'm still encountering issues. Additionally, I am unable to review this PR alongside the backend PR 906.

PR-2183

**Review by nyelmela — COMMENTED** [review]
I am unable to start `localhost:3000` due to errors and failed test suites. Despite following the steps mentioned, I am still encountering issues. Additionally, I am unable to review this PR alongside the backend PR #906. Thank you.

image

image

**Review by Pranay-ai — CHANGES_REQUESTED** [review]
Hey @ImzIssa 
I have tried testing the PR but i ran into few errors on the console here is the LOG file.
[localhost-1722615277507.log](https://github.com/user-attachments/files/16472831/localhost-1722615277507.log)

**Comment by Parth-tech** [comment]
Hey Imran,
After the dashboard loads up it goes blank. 

`Error:`
Screenshot 2024-08-15 at 6 22 19 PM
Screenshot 2024-08-15 at 6 22 27 PM
Screenshot 2024-08-15 at 12 00 04 AM

File: src/components/UserProfile/TeamsAndProjects/UserTeamsTable.jsx [hunk 2]
```diff
@@ -1,11 +1,14 @@
 import { React, useState } from 'react';
 import { Button, Input, Col, Tooltip } from 'reactstrap';
-import './TeamsAndProjects.css';
-import ToggleSwitch from '../UserProfileEdit/ToggleSwitch';
 import hasPermission from '../../../utils/permissions';
 import styles from './UserTeamsTable.css';
 import { boxStyle, boxStyleDark } from 'styles';
 import { connect } from 'react-redux';
+import { boxStyle } from 'styles';
```
Comment (Parth-tech) [inline src/components/UserProfile/TeamsAndProjects/UserTeamsTable.jsx:7]: You need to remove this line.

**Review by SMitta11 — CHANGES_REQUESTED** [review]
I ran this as per given instructions but on running frontend code getting error as boxStyle is imported twice, need to remove one import statement.
Screenshot:
Screen Shot 2024-08-17 at 1 11 51 AM
So I commented one of the import statement and ran again, there was no error this time but the app is showing unexpected results. It first gives the error and then page is blank if I refresh:
Screen Shot 2024-08-17 at 1 17 31 AM
Screen Shot 2024-08-17 at 1 19 40 AM

**Review by adityasure9 — APPROVED** [review]
I have tested the PR according to the provided instructions, and everything works as expected. The toggle functionality for invisibility permissions is functioning correctly, and I encountered no issues during testing. Great job on implementing these changes! Thank you!

image

**Review by oliviahyw — CHANGES_REQUESTED** [review]
Hi @ImzIssa I tested this PR and I encountered this issue: if I logged in as an Admin/Owner, my profile page is blank. But if I logged in as a Volunteer, I can see my profile page. Please check the attached screenshot.
Screenshot 2024-08-30 at 7 40 13 PM
Screenshot 2024-08-30 at 7 40 39 PM

**Review by ishangoel17 — COMMENTED** [review]
Verified the PR with the given branch code and observing blank screen issue while trying to navigate to the user profile page. Below are the screenshots for reference:
PR_Review_2183

**Review by Shreyav2000 — CHANGES_REQUESTED** [review]
Hi,

I followed the steps in your video, but when I log in as an Admin/Owner and navigate to any user profile (e.g., my test or volunteer user account) to assign a team, the page goes blank. I also tried viewing my own profile on the Admin page, and the console threw some errors.

Please refer to the video I'm uploading that shows the issue. Could you review and make the necessary changes?

Thanks!

https://github.com/user-attachments/assets/c5dbfc0c-4982-4a85-9b30-b74434f36d47

**Review by sam19981 — CHANGES_REQUESTED** [review]
Hi @ImzIssa 

Tried To test the changes but the website crashed as soon i tried to open the profile page for the admin and found the following errors.

Profile page crashes

so checked the code from the look of it you import is missing for the toggle switch in the 
[UserTeamsTable](src/components/UserProfile/TeamsAndProjects/UserTeamsTable.jsx).

**Review by vedantg24 — CHANGES_REQUESTED** [review]
Tested the PR, and faced some issues. When I login as Admin/Owner and go to my user profile or go to any other user's profile, the page goes blank, with errors in the console. However, if I log in as a volunteer, I can view my user profile. Attaching screenshots for the same.
Screenshot 2024-09-10
Screenshot 2024-09-10
Screenshot 2024-09-10

**Review by swaroop-2007 — CHANGES_REQUESTED** [review]
@ImzIssa,
I reviewed your PR and checked the code as well, but there seems to be an error with the functionality. Please refer to the below screenshots.
Screenshot (395)
Screenshot (396)

I guess there is some rendering issue. You may need to re-check the code. 
Thanks.

**Pull request closed without merging**